storage: Trigger a rebalance if diversity can be improved

a-robinson opened the pull request:
Fixes #17599, which we wanted to do early in the 1.2 release cycle.

I don't think this should be merged quite yet, due to the TODO on one of the new test cases that brings out bad diversity scoring behavior, but am sending out to maybe inspire some thought related to #17603.

I'd also like to add a few more test cases to this once the diversity score changes have settled down.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/19489)
<!-- Reviewable:end -->

BramGruneir commented:
This change looks good, and we should be triggering rebalances based on diversity improvements.  But I'll leave it up to you for now to improve testing of it before merging.  I think we might want to create a few new tests similar to Example_rebalancing, that deal with a few different common scenarios with localities included.  This way, when we do change diversity score calculations, we can get an immediate sense of how that change affects common cluster setups.  They should probably move to a new test file too, since allocator_test is already huge.

a6802739 commented on pkg/storage/allocator_scorer.go:
```diff
@@ -552,13 +552,30 @@ func shouldRebalance(
+			continue
+		}
+		otherScore := rebalanceToDiversityScore(desc, existingNodeLocalities)
+		if otherScore > diversityScore {
```
@a-robinson , This is great!

I have some confusion here, how could we gurantee we could achieve better locality for this range if `otherScore > diversityScore`?

a6802739 commented on pkg/storage/allocator_scorer.go:
```diff
@@ -552,13 +552,30 @@ func shouldRebalance(
+			continue
+		}
+		otherScore := rebalanceToDiversityScore(desc, existingNodeLocalities)
+		if otherScore > diversityScore {
```
@a-robinson , This is great!

I have some confusion here, how could we gurantee we could achieve better locality for this range if otherScore > diversityScore?

a-robinson replied:
I've added a test that ensures the method used to compute the current diversity and the method used to compute what the diversity would be if we added desc and removed store are directly comparable. I believe that resolves this, but do you still see anything wrong with it?

a6802739 commented on pkg/storage/allocator_scorer_test.go:
```diff
@@ -489,6 +490,147 @@ func TestConstraintCheck(t *testing.T) {
+			s:               newStore(1, locUS),
+			sl:              sl3by3,
+			existingNodeIDs: []roachpb.NodeID{1, 4, 5},
+			expected:        false,
```
Why here `expected` should be false? From my understanding, we could rebalance to `locAS` to achieve better locality diversity.

a6802739 replied:
Yeah, I see now. Thank you for your explanation.

a6802739 commented on pkg/storage/allocator_scorer_test.go:
```diff
@@ -489,6 +490,147 @@ func TestConstraintCheck(t *testing.T) {
+			s:               newStore(1, locUS),
+			sl:              sl4by3,
+			existingNodeIDs: []roachpb.NodeID{1, 2, 4, 7, 10},
+			expected:        true,
```
How about we check if we rebalance to a target store, it will improve the overall diversity for this range?

In this methods, we should first get to know which store we are going to remove, and then calculates the overall diversityScore for each target rebalance store, and compare the diversityScore with the original overall diversity of the range. just like #19486.

a6802739 reviewed (commented):
@a-robinson  This is Great!

a-robinson committed:
storage: Trigger a rebalance if diversity can be improved

Fixes #17599, which we wanted to do early in the 1.2 release cycle.

Release note (core change): Proactively rebalance data if diversity of
the localities that a given range is located on can be improved.

a-robinson reviewed (commented):
Thanks for the reviews! This should be good to go now that #19486 is in.

a6802739 commented on pkg/storage/allocator_scorer.go:
```diff
@@ -697,6 +713,30 @@ func constraintCheck(store roachpb.StoreDescriptor, constraints config.Constrain
+		for n2, l2 := range existingNodeLocalities {
+			// Only compare pairs of replicas where s2 > s1 to avoid computing the
+			// diversity score between each pair of localities twice.
+			if n2 <= n1 {
```
Great idea!

a6802739 commented on pkg/storage/allocator_scorer_test.go:
```diff
@@ -757,6 +903,64 @@ func TestRemovalDiversityScore(t *testing.T) {
+	// the same results for the same configurations, enabling their results
+	// to be directly compared with each other. The same is not true for
+	// diversityAllocateScore and diversityRemovalScore as of their initial
+	// creation or else we would test them here as well.
```
I could understand if the remove target and the rebalance target are same, `rangeDiversityScore` will be same with `diversityRebalanceFromScore`, but I think maybe we should have a test that we could achieve better diversity score for`diversityRebalanceFromScore` if we remove one replica and rebalance to one target comparing with `rangeDiversityScore`.

a-robinson replied:
I don't think I understand exactly what you're suggesting. It sounds to me like you're suggesting something along the lines of what I've already added in `TestShouldRebalanceDiversity`, which is testing that new code in `shouldRebalance` is able to use `rangeDiversityScore` and `diversityRebalanceFromScore` to properly determine when a rebalance would be beneficial for diversity. Could you explain in more detail what you're suggesting and what code it would be testing?

a6802739 replied:
Hmm, Sorry for my bad expression. What I mean is that maybe we could add a test. The test will display the result of both functions, and literally the result of `diversityRebalanceFromScore` is obviously bigger than that of `rangeDiversityScore`.
`TestShouldRebalanceDiversity` proves that the result of which function is bigger but doesn't display the result of both functions.
Maybe it's unnecessary, I just think it's a better way for other guys to understand.

a-robinson replied:
Ok, I see now. That would help with understanding, but I don't think it provides any more code coverage. I'm going to skip it for now, but it would be a good test to have.

a6802739 commented on pkg/storage/allocator_scorer.go:
```diff
@@ -550,14 +550,30 @@ func shouldRebalance(
+		if !preexistingReplicaCheck(desc.Node.NodeID, rangeInfo.Desc.Replicas) {
+			continue
+		}
+		otherScore := diversityRebalanceFromScore(desc, store.Node.NodeID, existingNodeLocalities)
```
I think It's quite obvious now.

a6802739 reviewed (commented):
LGTM!
